packages/php-wasm/web/src/lib/web-php.ts

```diff
@@ -44,6 +48,560 @@ const fakeWebsocket = () => {
+	//cert.setSubject(csr.subject.attributes);
+	cert.setIssuer(attrs);
+	cert.setExtensions([
+		{
+			name: 'basicConstraints',
+			cA: true,
+		},
+		{
+			name: 'keyUsage',
+			keyCertSign: true,
+			digitalSignature: true,
+			nonRepudiation: true,
+			keyEncipherment: true,
+			dataEncipherment: true,
+		},
+		{
+			name: 'extKeyUsage',
+			serverAuth: true,
+			clientAuth: true,
+			codeSigning: true,
+			emailProtection: true,
+			timeStamping: true,
+		},
+		{
+			name: 'nsCertType',
+			client: true,
+			server: true,
+			email: true,
+			objsign: true,
+			sslCA: true,
+			emailCA: true,
+			objCA: true,
+		},
+		{
+			name: 'subjectAltName',
+			altNames: [
+				{
+					type: 6, // URI
+					value: '*.org',
+				},
+				{
+					type: 7, // IP
+					ip: '127.0.0.1',
+				},
+			],
+		},
+		{
+			name: 'subjectKeyIdentifier',
+		},
+	]);
+	// FIXME: add subjectKeyIdentifier extension
+	// FIXME: add authorityKeyIdentifier extension
+	cert.publicKey = keys.publicKey;
+
+	// self-sign certificate
+	cert.sign(keys.privateKey);
+
+	// console.log('certificate created for \"' + forge.pki.certificateToPem(cert) + '\": \n' + forge.pki.privateKeyToPem(keys.privateKey));
+	return { certificate: cert, privateKey: keys.privateKey };
+}
+
+// Introduce artificial delay to test the race
+// condition when initializing the Playground client.
+// for (let i = 0; i < 2; i++) {
+// 	createCertificate();
+// }
+
+// @TODO: Create these certificates in FetchWebsocketConstructor based on the requested host
+const { certificate, privateKey } = createCertificate({
+	commonName: 'downloads.wordpress.org', //wsUrl.searchParams.get('host') || ''
+});
+const CAPair = { certificate, privateKey };
+// const CAPair = createCertificate({
+// 	commonName: ''
+// });
+export const CAPem = forge.pki.certificateToPem(CAPair.certificate);
+/**
+ * Websocket that buffers the received bytes and translates them into
+ * a fetch() call.
+ */
+const fetchingWebsocket = (phpModuleArgs: EmscriptenOptions = {}) => {
+	return {
+		websocket: {
+			...(phpModuleArgs['websocket'] || {}),
+			url: (_: any, host: string, port: string) => {
+				const query = new URLSearchParams({
+					host,
+					port,
+				}).toString();
+				return `ws://playground.internal/?${query}`;
+			},
+			subprotocol: 'binary',
+			decorator: () => {
+				return class FetchWebsocketConstructor {
+					CONNECTING = 0;
+					OPEN = 1;
+					CLOSING = 2;
+					CLOSED = 3;
+					readyState = this.CONNECTING;
+					binaryType = 'blob';
+					bufferedAmount = 0;
+					extensions = '';
+					protocol = 'ws';
+					host = '';
+					port = 0;
+					listeners = new Map<string, any>();
+					sslServer: any;
+					isPlaintext: boolean | null = null;
+
+					constructor(public url: string, public options: string[]) {
+						const wsUrl = new URL(url);
+						this.host = wsUrl.searchParams.get('host')!;
+						this.port = parseInt(
+							wsUrl.searchParams.get('port')!,
+							10
+						);
+
+						const ws = this;
```

Comment:
copy/paste or AI code?

Reply:
Quick&dirty prototype, `ws` is used in a `function` in line 316 below where `this` refers to a different object.